R-package/R/lgb.Dataset.R

```diff
@@ -13,7 +13,12 @@ Dataset <- R6::R6Class(
       if (!lgb.is.null.handle(x = private$handle)) {
 
         # Freeing up handle
-        lgb.call(fun_name = "LGBM_DatasetFree_R", ret = NULL, private$handle)
+        call_state <- 0L
```

Comment:
Even though error-handling has now been moved into the C++ side (#4163), as of this PR passing in `call_state` is still needed. The functions in https://github.com/microsoft/LightGBM/blob/b6c71e5e941b885b1a5169d460c430df9d392fa9/R-package/src/lightgbm_R.h has a return type `LGBM_SE`, which is the LightGBM equivalent of R's built-in `SEXP` type.

If you change those routines to all be `void` (since they modify their inputs by reference and don't technically have to refer to anything), errors are thrown about not being able to convert a null to `SEXP`.

So the `call_state` thing is a simple way to allocate an `LGBM_SE` (an R integer, in this case) that can then be returned.

As part of #3016 we'll be able to remove these unnecessary `call_state` things, but I believe they're harmless for now.